src/data/articles/scala-3-the-evolution-of-a-macro.mdx

```diff
@@ -0,0 +1,1829 @@
+  - Metadata fields to be added the tracing call, meant to serve as a list of key/value entries (3)
+  - The actual code being traced, again by-name, as we want to suspend its evaluation to inject tracing before/after it runs (4)
+  - A tracer, to be filled in by the compiler so as to not bother the users of the tracing library (5)
+- In the implementation itself:
+  - We first check whether tracing is enabled for the current level (6)
+  - Only then we invoke our tracing call with all the metadata (7)
+  - We wrap everything with a `try` so as to be able to catch failures and signal them to the tracer (8)
+  - We run the body (9)
+  - If we finished successfully, we notify the tracer and add the result as metadata[^avoidToString] (10)
+  - And we return the result to the caller (11)
+  - If anything goes wrong, we catch all non-fatal exceptions (12)
+  - And notify the tracer[^againToString]  (13)
+  - And re-throw the exception back to the caller (14)
+  - If tracing is disabled we run the body without any instrumentation (15)
+
+[^avoidToString]: This is for illustration purposes only, using `toString` like this is probably not what you want in a generic context.
+[^againToString]: Again, `toString` for illustration only.
+
+And here's a usage example exercising this code:
+
+```scala wrap=false
+class ClassToBeTraced0:
+  import Baseline.*
+
+  // 1
+  private given tracer: Tracer = Tracer.make[ClassToBeTraced0]
+
+  // 2
+  def theMethodToBeTraced(arg1: String,
+                          arg2: Int,
+                          arg3: SomeData,
+                          arg4: UntraceableData): String =
+    trace0(
+      level = TraceLevel.Info,
+      id = "theMethodToBeTraced", // 3
+      "arg1" -> arg1.toString,
+      "arg2" -> arg2.toString,
+      "arg3" -> arg3.toString,
+      "arg4" -> arg4.toString):
+
+        if arg2 > 2 then "hello" // 4
+        else "bye"
+```
+
+In this example:
+- We initialize a given tracer (1)
+- `theMethodToBeTraced` is the method that we actually want to be traced (2)
+- To trace the invocation of this method we'll use the method name as the ID and add all the arguments we got as metadata (3)
+- And then we pass in the actual body of the method (4).
+
+This code works, but one concern about it is that it might lose on performance as we are wrapping arguments in by-name parameters. Using `trace0` might be slower than writing the `if/else` manually every time.
+
+Ideally we would want to avoid the `if/else` boilerplate but also avoid the performance cost. Of course you might not care about performance in this context, and then this code is perfectly fine, but imagine that this is a generic library code and you want your users to be able to use it "for free".
+
+Let's see how we can improve on this.
+
+## With Inlining
+
+Scala 3 has a builtin way to avoid the performance cost incurred by using by-name arguments: `inline`s.
+You can read more about them [here](scala-3-inlines).
```

Comment:
I think the link should be `/articles/scala-3-inlines`, in any case make sure the links work

Reply:
The relative links work as is. Do you prefer to use absolute links?